Mathlib/Data/Matroid/Map.lean

```diff
@@ -0,0 +1,669 @@
+
+## Main definitions
+
+In the definitions below, `M` and `N` are matroids on `α` and `β` respectively.
+
+* For `f : α → β`, `Matroid.comap N f` is the matroid on `α` with ground set `f ⁻¹' N.E`
+  in which each `I` is independent if and only if `f` is injective on `I` and
+  `f '' I` is independent in `N`.
+  (For each nonloop `x` of `N`, the set `f ⁻¹' {x}` is a parallel class of `N.comap f`)
+
+* `Matroid.comapOn N f E` is the restriction of `N.comap f` to `E` for some `E : Set α`.
+
+* For an embedding `f : M.E ↪ β` defined on the subtype `↑M.E`,
+  `Matroid.mapSetEmbedding M f` is the matroid on `β` with ground set `range f`
+  whose independent sets are the images of those in `M`. This matroid is isomorphic to `M`.
+
+* For a function `f : α → β` and a proof `hf` that `f` is injective on `M.E`,
+  `Matroid.map f hf` is the matroid on `β` with ground set `f '' M.E`
+  whose independent sets are the images of those in `M`. This matroid is isomorphic to `M`,
+  and does not depend on the values `f` takes outside `M.E`.
+
+* `Matroid.mapEmbedding f` is a version of `Matroid.map` where `f : α ↪ β` is a bundled embedding.
+  It is defined separately because the global injectivity of `f` gives some nicer `simp` lemmas.
+
+* `Matroid.mapEquiv f` is a version of `Matroid.map` where `f : α ≃ β` is a bundled equivalence.
+  It is defined separately because we get even nicer `simp` lemmas.
+
+* `Matroid.mapSetEquiv f` is a version of `Matroid.map` where `f : M.E ≃ E` is an equivalence on
+  subtypes. It gives a matroid on `β` with ground set `E`.
+
+* For `X : Set α`, `Matroid.restrictSubtype M X` is the `Matroid X` with ground set
+  `univ : Set X` that is isomorphic to `M ↾ X`.
+
+## Implementation details
+
+The definition of `comap` is the only place where we need to actually define a matroid from scratch.
+After `comap` is defined, we can define `map` and its variants indirectly in terms of `comap`.
+
+If `f : α → β` is injective on `M.E`, the independent sets of `M.map f hf` are the images of
+the independent set of `M`; i.e. `(M.map f hf).Indep I ↔ ∃ I₀, M.Indep I₀ ∧ I = f '' I₀`.
+But if `f` is globally injective, we can phrase this more directly;
+indeed, `(M.map f _).Indep I ↔ M.Indep (f ⁻¹' I) ∧ I ⊆ range f`.
+If `f` is an equivalence we have `(M.map f _).Indep I ↔ M.Indep (f.symm '' I)`.
+In order that these stronger statements can be `@[simp]`,
+we define `mapEmbedding` and `mapEquiv` separately from `map`.
+
+## Notes
+
+For finite matroids, both maps and comaps are a special case of a construction of
+Perfect (1969) in which a matroid structure can be transported across an arbitrary
+bipartite graph that may not correspond to a function at all [See Oxley, Thm 11.2.12].
+It would have been nice to use this more general construction as a basis for the definition
+of both `Matroid.map` and `Matroid.comap`.
+
+Unfortunately, we can't do this, because the construction doesn't extend to infinite matroids.
```

Comment:
I guess you could have a `mapFinite` or something like that. But maybe that's too many variants?

Reply:
I have a student working on this at the moment. It has quite a different flavour from what is currently here, since it actually has some mathematical content. I'll put it as a todo.